File: course_schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        prereqMap = {i: [] for i in range(numCourses)}

        for course, prereq in prerequisites:
            prereqMap[course].append(prereq)

        visited = set()

        def dfs(course):
            if course in visited:
                return False

            if prereqMap[course] == []:
                return True

            visited.add(course)

            for prereq in prereqMap[course]:
                if not dfs(prereq):
                    return False

            visited.remove(course)
            prereqMap[course] = []
            return True

        for c in range(numCourses):
            if not dfs(c):
                return False

        return True

File: courseSchedule.js
class Solution {
  /**
   * @param {number} numCourses
   * @param {number[][]} prerequisites
   * @return {boolean}
   */
  canFinish(numCourses, prerequisites) {
    const preMap = {};

    for (let i = 0; i < numCourses; i++) {
      preMap[i] = [];
    }

    for (const [course, prereq] of prerequisites) {
      preMap[course].push(prereq);
    }

    const visited = new Set();

    const dfs = (course) => {
      if (visited.has(course)) return false;
      if (preMap[course].length === 0) return true;

      visited.add(course);

      for (let prereq of preMap[course]) {
        if (!dfs(prereq)) {
          return false;
        }
      }
      visited.delete(course);
      preMap[course] = [];
      return true;
    };

    for (let c = 0; c < numCourses; c++) {
      if (!dfs(c)) return false;
    }

    return true;
  }
}
